pinot-plugins/pinot-segment-writer/pinot-segment-writer-file-based/src/main/java/org/apache/pinot/plugin/segmentwriter/filebased/FileBasedSegmentWriter.java

```diff
@@ -0,0 +1,193 @@
+  private Set<String> _fieldsToRead;
+  private RecordTransformer _recordTransformer;
+
+  private File _stagingDir;
+  private File _bufferFile;
+
+  private org.apache.avro.Schema _avroSchema;
+  private DataFileWriter<GenericData.Record> _recordWriter;
+  private GenericData.Record _reusableRecord;
+
+  @Override
+  public void init(TableConfig tableConfig, Schema schema)
+      throws Exception {
+    _tableConfig = tableConfig;
+    _tableNameWithType = _tableConfig.getTableName();
+
+    Preconditions.checkState(
+        _tableConfig.getIngestionConfig() != null && _tableConfig.getIngestionConfig().getBatchIngestionConfig() != null
+            && CollectionUtils
+            .isNotEmpty(_tableConfig.getIngestionConfig().getBatchIngestionConfig().getBatchConfigMaps()),
+        "Must provide ingestionConfig->batchIngestionConfig->batchConfigMaps in tableConfig for table: %s",
+        _tableNameWithType);
+    _batchIngestionConfig = _tableConfig.getIngestionConfig().getBatchIngestionConfig();
+    _batchConfig = new BatchConfig(_tableNameWithType, _batchIngestionConfig.getBatchConfigMaps().get(0));
+
+    Preconditions.checkState(StringUtils.isNotBlank(_batchConfig.getOutputDirURI()),
+        "Must provide: %s in batchConfigs for table: %s", BatchConfigProperties.OUTPUT_DIR_URI, _tableNameWithType);
+    _outputDirURI = _batchConfig.getOutputDirURI();
+    Files.createDirectories(Paths.get(_outputDirURI));
+
+    _schema = schema;
+    _fieldsToRead = _schema.getColumnNames();
+    _recordTransformer = CompositeTransformer.getDefaultTransformer(_tableConfig, _schema);
+    _avroSchema = SegmentProcessorAvroUtils.convertPinotSchemaToAvroSchema(_schema);
+    _reusableRecord = new GenericData.Record(_avroSchema);
+
+    // Create tmp dir
+    _stagingDir = new File(FileUtils.getTempDirectory(),
+        String.format("segment_writer_staging_%s_%d", _tableNameWithType, System.currentTimeMillis()));
+    Preconditions.checkState(_stagingDir.mkdirs(), "Failed to create staging dir: %s", _stagingDir.getAbsolutePath());
+
+    // Create buffer file
+    File bufferDir = new File(_stagingDir, "buffer_dir");
+    Preconditions.checkState(bufferDir.mkdirs(), "Failed to create buffer_dir: %s", bufferDir.getAbsolutePath());
+    _bufferFile = new File(bufferDir, "buffer_file");
+    resetBuffer();
+    LOGGER.info("Initialized {} for table: {}", FileBasedSegmentWriter.class.getName(), _tableNameWithType);
+  }
+
+  private void resetBuffer()
+      throws IOException {
+    FileUtils.deleteQuietly(_bufferFile);
+    _recordWriter = new DataFileWriter<>(new GenericDatumWriter<>(_avroSchema));
+    _recordWriter.create(_avroSchema, _bufferFile);
+  }
+
+  @Override
+  public void collect(GenericRow row)
+      throws IOException {
+    GenericRow transform = _recordTransformer.transform(row);
+    SegmentProcessorAvroUtils.convertGenericRowToAvroRecord(transform, _reusableRecord, _fieldsToRead);
+    _recordWriter.append(_reusableRecord);
```

Comment:
shall we block this call when `flush` is in progress?

Reply:
added the NotThreadSafe annotation. Do we still need to block it?  Do you anticipate the caller would invoke flush in a separate executor thread?

Reply:
With annotation I think it's good. 

Also, `flush` calls `_recordWriter.close()`, which means the caller cannot call `collect` or `flush` again after the call on `flush`, right?

Reply:
Caller can call both `collect` and `flush` again after `flush`, because of the `resetBuffer` call right before returning the `URI` in the `flush` call.
Check out the `FileBasedSegmentWriterTest#testSegmentWriter` or the `SegmentWriterUploaderIntegrationTest` which uses same `SegmentWriter` and calls collect again after flush. And then `FileBasedSegmentWriterTest#testEmptySegment` which calls flush even without any records collected (will work for flush immediately followed by flush).

Reply:
Got it. Thanks.